plugins/pulp_rpm/plugins/importers/yum/sync.py

```diff
@@ -127,70 +168,91 @@ def run(self):
+            # we delete below
+            self.tmp_dir = tempfile.mkdtemp(dir=self.working_dir)
+            url_count += 1
+            try:
+                with self.update_state(self.progress_status['metadata']):
+                    metadata_files = self.check_metadata(url)
+                    metadata_files = self.get_metadata(metadata_files)
+
+                    # Save the default checksum from the metadata
+                    self.save_default_metadata_checksum_on_repo(metadata_files)
+
+                with self.update_state(self.content_report) as skip:
+                    if not (skip or self.skip_repomd_steps):
+                        self.update_content(metadata_files, url)
+
+                _logger.info(_('Downloading additional units.'))
+
+                with self.update_state(self.distribution_report, models.Distribution.TYPE) as skip:
+                    if not skip:
+                        treeinfo.sync(self.sync_conduit, url, self.tmp_dir,
+                                      self.nectar_config, self.distribution_report,
+                                      self.set_progress)
+
+                with self.update_state(self.progress_status['errata'], models.Errata.TYPE) as skip:
+                    if not (skip or self.skip_repomd_steps):
+                        self.get_errata(metadata_files)
+
+                with self.update_state(self.progress_status['comps']) as skip:
+                    if not (skip or self.skip_repomd_steps):
+                        self.get_comps_file_units(metadata_files, group.process_group_element,
+                                                  group.GROUP_TAG)
+                        self.get_comps_file_units(metadata_files, group.process_category_element,
+                                                  group.CATEGORY_TAG)
+                        self.get_comps_file_units(metadata_files, group.process_environment_element,
+                                                  group.ENVIRONMENT_TAG)
+
+            except CancelException:
+                report = self.sync_conduit.build_cancel_report(self._progress_summary,
+                                                               self.progress_status)
+                report.canceled_flag = True
+                return report
+
+            except Exception, e:
+                # Check if the caught exception indicates that the mirror is bad.
+                # Try next mirror in the list without raising the exception.
+                # In case it was the last mirror in the list, raise the exception.
+                # In case other exceptions were caught that are not related to the state of the
+                # mirror, raise the exception immediately and do not iterate throught the rest
+                # of mirrors.
+                bad_mirror_exceptions = [error_codes.RPM1004, error_codes.RPM1006]
+                if isinstance(e, PulpCodedException) and \
```

Comment:
Rather than catching all Exceptions and performing an isinstance() check, you should catch PulpCodedExceptions. Better yet, you should catch the specific Exceptions that correspond to 1004 or 1006 here.